query/src/main/java/org/infinispan/query/dsl/embedded/impl/JPAFilterIndexingServiceProvider.java

```diff
@@ -0,0 +1,324 @@
+      cacheNotifier.getListenerCollectionForAnnotation(CacheEntryCreated.class).removeAll(filteringInvocations.values());
+      cacheNotifier.getListenerCollectionForAnnotation(CacheEntryInvalidated.class).removeAll(filteringInvocations.values());
+      cacheNotifier.getListenerCollectionForAnnotation(CacheEntryLoaded.class).removeAll(filteringInvocations.values());
+      cacheNotifier.getListenerCollectionForAnnotation(CacheEntryModified.class).removeAll(filteringInvocations.values());
+      cacheNotifier.getListenerCollectionForAnnotation(CacheEntryPassivated.class).removeAll(filteringInvocations.values());
+      cacheNotifier.getListenerCollectionForAnnotation(CacheEntryRemoved.class).removeAll(filteringInvocations.values());
+      cacheNotifier.getListenerCollectionForAnnotation(CacheEntryVisited.class).removeAll(filteringInvocations.values());
+      cacheNotifier.getListenerCollectionForAnnotation(CacheEntryEvicted.class).removeAll(filteringInvocations.values());
+      filteringInvocations.clear();
+   }
+
+   @Override
+   public boolean supportsFilter(IndexedFilter<?, ?, ?> indexedFilter) {
+      return indexedFilter instanceof JPACacheEventFilterConverter;
+   }
+
+   @Override
+   public <K, V> DelegatingCacheEntryListenerInvocation<K, V> interceptListenerInvocation(CacheEntryListenerInvocation<K, V> invocation) {
+      return new DelegatingCacheEntryListenerInvocationImpl<K, V>(invocation);
+   }
+
+   @Override
+   public <K, V> void registerListenerInvocations(boolean isClustered, boolean isPrimaryOnly, boolean filterAndConvert,
+                                                  IndexedFilter<?, ?, ?> indexedFilter,
+                                                  Map<Class<? extends Annotation>, List<DelegatingCacheEntryListenerInvocation<K, V>>> listeners) {
+      JPAFilterAndConverter filter = ((JPACacheEventFilterConverter) indexedFilter).filterAndConverter;
+      Matcher matcher = filter.getMatcher();
+      addFilteringInvocationForMatcher(matcher);
+      Event.Type[] eventTypes = new Event.Type[listeners.keySet().size()];
+      int i = 0;
+      for (Class<? extends Annotation> annotation : listeners.keySet()) {
+         eventTypes[i++] = getEventType(annotation);
+      }
+      Callback<K, V> callback = new Callback<K, V>(matcher, isClustered, isPrimaryOnly, filterAndConvert, listeners);
+      callback.subscription = matcher.registerFilter(filter.getJPAQuery(), callback, eventTypes);
+   }
+
+   /**
+    * Obtains the event type that corresponds to the given event annotation.
+    *
+    * @param annotation a CacheEntryXXX annotation
+    * @return the event type or {@code null} if the given annotation is not supported
+    */
+   private Event.Type getEventType(Class<? extends Annotation> annotation) {
+      if (annotation == CacheEntryActivated.class) return Event.Type.CACHE_ENTRY_ACTIVATED;
```

Comment:
I think it would be interesting to see, but not being picky about it.  But I would say we should move CacheEntryCreated, CacheEntryModified, CacheEntryRemoved to be first - most likely these will be the events people listen to the most.